### PR TITLE
introduce include to load sub-compose projects as dependencies

### PR DESCRIPTION
**Pull request by ndeloof**
This adds support for `include` attribute introduced in the compose-specification by https://github.com/compose-spec/compose-spec/pull/363

This supports including compose sub-projects, comparable to language dependency management

usage:
```yaml
include:
   - ../commons/compose.yaml
   - ../another-project/compose.yaml

services:
   foo:
      depends_on:
         - imported-service
```

required projects are resolved based on their own project directory to avoid relative path hell.

### CODE REVIEW

File: loader/include.go [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+var transformRequireConfig TransformerFunc = func(data interface{}) (interface{}, error) {
+	switch value := data.(type) {
+	case string:
+		return map[string]interface{}{"path": value}, nil
+	case map[string]interface{}:
+		return value, nil
+	default:
+		return data, errors.Errorf("invalid type %T for `require` configuration", value)
```
Comment (milas) [inline loader/include.go:42]:
```suggestion
		return data, errors.Errorf("invalid type %T for `include` configuration", value)
```

**Review by milas — APPROVED** [review]
LGTM

File: loader/include.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+)
+
+// LoadIncludeConfig parse the require config from raw yaml
+func LoadIncludeConfig(source []interface{}) ([]types.IncludeConfig, error) {
+	var requires []types.IncludeConfig
+	err := Transform(source, &requires)
+	return requires, err
+}
+
+var transformIncludeConfig TransformerFunc = func(data interface{}) (interface{}, error) {
```
Comment (glours) [inline loader/include.go:35]: nit: can't we use generics here?

File: loader/loader.go [hunk 3]
```diff
@@ -262,7 +304,7 @@ func Load(configDetails types.ConfigDetails, options ...func(*Options)) (*types.
 	}
 
 	if opts.ResolvePaths {
-		err = ResolveRelativePaths(project)
+		err := ResolveRelativePaths(project)
 		if err != nil {
 			return nil, err
 		}
```
Comment (glours) [inline loader/loader.go:310]: nit: could also be apply to the following `if !opts.` blocks
```suggestion
		if err := ResolveRelativePaths(project); err != nil {
			return nil, err
		}
```

File: loader/testdata/compose-include.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+include:
+  - compose-include.yaml
+
+services:
+  foo:
+    image: foo
```
Comment (glours) [inline loader/testdata/compose-include.yaml:6]: nit: new line

**Review by glours — APPROVED** [review]
Sounds good to me, you just need to fix the conflict before merging

File: loader/loader.go (hunk 3 above)
Comment (ndeloof) [reply loader/loader.go:310]: I personally consider those to have lower readability for the tiny benefit of a single line being saved

File: loader/include.go (hunk 2 above)
Comment (ndeloof) [reply loader/include.go:35]: this signature isn't a "lack of generics": those manage the parsed yaml nodes, which are actual `interface{}`

File: loader/loader.go (hunk 3 above)
Comment (glours) [reply loader/loader.go:310]: As you wish, I don't have a strong opinion about this 😉

**Comment by vasilvestre** [comment]
Could it work with URL later ?

**Comment by ndeloof** [comment]
@vasilvestre this is indeed something we consider for future, also see https://github.com/compose-spec/compose-spec/issues/288

**Comment by VladaPetrovic** [comment]
The functionality to include the remote file would be a great addition to this feature 🚀 
It was already proposed in the original [idea](https://github.com/docker/compose/issues/318#issue-37799766)

for example:
```
include:
https://localhost:8000/some-service/compose.yaml
...
```

We could leverage this feature to re-use Compose files across teams. Currently, we’re cloning an additional repository solely to include another Compose file, which feels a bit cumbersome. 🤔

The mentioned https://github.com/compose-spec/compose-spec/issues/288 seems like a broader feature with a much larger scope. While it could also address this problem, it approaches it from a different perspective.

File: dotenv/godotenv_test.go [hunk 5]
```diff
@@ -3,21 +3,20 @@ package dotenv
...
-	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"gotest.tools/v3/assert"
 )
 
 var noopPresets = make(map[string]string)
 
 func parseAndCompare(t *testing.T, rawEnvLine string, expectedKey string, expectedValue string) {
 	t.Helper()
 	env, err := Parse(strings.NewReader(rawEnvLine))
```
Comment (colliniivan948-ops) [inline dotenv/godotenv_test.go:18]: ###

**Pull request merged**